Support DSR with IPIP in CNI mode

Jiang1155 opened the pull request:
<!-- Support DSR with IPIP in CNI mode -->

```release-note
Add support for DSR using IPIP. The switches will forward the IPIP packets via a fast path.
```

pchaigno commented on bpf/include/bpf/ctx/xdp.h:
```diff
@@ -209,8 +209,10 @@ ctx_adjust_hroom(struct xdp_md *ctx, const __s32 len_diff, const __u32 mode,
+	build_bug_on(len_diff == 0 || len_diff >= 64);
 	build_bug_on(mode != BPF_ADJ_ROOM_NET);
+	if (len_diff < 0)
+		return xdp_adjust_head(ctx, 0 - len_diff);
```
Is this needed? You only call `ctx_adjust_hroom` from `handle_dsr_v4` and that function is only called from bpf_lxc, so the XDP implementation shouldn't be needed for `len_diff` < 0, no?

Jiang1155 replied:
This is needed because in handle_dsr_v4, I added a new line to call ctx_adjust_hroom with a negative offset.

pchaigno replied:
Ok, but `handle_dsr_v4` is only called from `bpf_lxc.c` which is never loaded in an XDP context (only tc). So I don't understand why changes to this limited XDP implementation are needed. `handle_dsr_v4` will use the tc implementation of `ctx_adjust_hroom` which simply relies on the `bpf_skb_adjust_room` helper.

Jiang1155 replied:
I got your point now. I will remove this code and test it again. Thanks.

Jiang1155 replied:
Sorry for the late reply. I was fixing some other serious bugs before this one. I just tried to remove it and it is like what you said.  I will revert it in the next version. Thanks.

pchaigno commented on bpf/include/bpf/ctx/xdp.h:
```diff
@@ -210,7 +210,7 @@ ctx_adjust_hroom(struct xdp_md *ctx, const __s32 len_diff, const __u32 mode,
 
 	build_bug_on(len_diff == 0 || len_diff >= 64);
-	build_bug_on(mode != BPF_ADJ_ROOM_NET);
+	build_bug_on((len_diff > 0) &&(mode != BPF_ADJ_ROOM_NET));
```
This should be in the previous commit, no?

Jiang1155 replied:
yes. I will fix it.

pchaigno commented on bpf/lib/conntrack.h:
```diff
@@ -499,7 +499,9 @@ ipv4_extract_tuple(struct __ctx_buff *ctx, struct ipv4_ct_tuple *tuple,
 	if (unlikely(tuple->nexthdr != IPPROTO_TCP &&
-		     tuple->nexthdr != IPPROTO_UDP))
+		     tuple->nexthdr != IPPROTO_UDP &&
+		     tuple->nexthdr != IPPROTO_IPIP))
```
Why is this change needed?

Jiang1155 replied:
Without this change, the IPIP packets will be dropped  before handle_dsr_v4.

pchaigno replied:
Did you test that? I see `ipv4_extract_tuple` is called from `extract_tuple4`, which is itself called only from `ctx_redirect_to_proxy_first`. `ctx_redirect_to_proxy_first` is called only from `bpf_host`'s `to-host` section, only in case of a packet going to the proxy. Am I missing something?

Jiang1155 replied:
I see. I did not test this change alone, but tested it with another change. I will remove it and test again. Thanks.
btw, I saw other comments about adding new test, updating doc, refactoring a function etc. I will try to address all of them.

pchaigno replied:
Awesome! Don't hesitate to ask for help if needed. I'm pchaigno on Slack as well.

pchaigno commented on bpf/lib/conntrack.h:
```diff
@@ -615,7 +617,9 @@ ct_extract_ports4(struct __ctx_buff *ctx, int off, int dir,
 		break;
-
+	case IPPROTO_IPIP:
+		off = off + 20;
```
We need to check that the outer IP header isn't >20B (doesn't carry IPv4 options).

Also, let's use `sizeof(struct iphdr)` here.

Jiang1155 replied:
If the outer IP header is longer than 20B, what should we do? Drop it? I think another way is to check the outer IP header length and use it to replace 20B. What do you think?

I will use sizeof(struct iphdr), thanks.

pchaigno replied:
> I think another way is to check the outer IP header length and use it to replace 20B. What do you think?

Yes, I think that's the right approach.

pchaigno commented on bpf/lib/conntrack.h:
```diff
@@ -725,7 +729,26 @@ static __always_inline int ct_lookup4(const void *map,
+					offsetof(struct iphdr, protocol), &protocol, 1) < 0)
+				return DROP_CT_INVALID_HDR;
 
+			off = off + iphdrlen;
```
We should use `ipv4_hdrlen()` here to retrieve the actual size of the inner IP header.

Jiang1155 replied:
sure.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
 #endif /* DSR_ENCAP_MODE */
 
-static __always_inline int handle_dsr_v4(struct __ctx_buff *ctx, bool *dsr)
+static __always_inline int handle_dsr_v4(struct __ctx_buff *ctx, bool *dsr, int ct)
```
We recently introduced an enumeration for the conntrack states:
```suggestion
static __always_inline int
handle_dsr_v4(struct __ctx_buff *ctx, bool *dsr, enum ct_status ct)
```

Jiang1155 replied:
fixed

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+					return DROP_CT_INVALID_HDR;
+
+				if (snat_v4_create_dsr(ctx, vip, dport) < 0)
+					return DROP_INVALID;
```
Could you refactor this function to share this call with the `DSR_ENCAP_NONE` case? You probably will need to rework a bit the existing code in a preparatory commit.

Jiang1155 replied:
I modified this part in the newer version. But did not share the call to snat_v4_create_dsr(). With the new logic, to share the call is a little bit complicated. We can chat more if you still think it is better to share the call. Thanks.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+		__be32 sip __maybe_unused;
+		__be32 sum __maybe_unused;
+		__be16 dport __maybe_unused = 0;
+		int l4_off __maybe_unused;
```
All the `__maybe_unused` seem unnecessary here. Maybe they are from a previous version of the changes?

Jiang1155 replied:
fixed. thanks

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+					    0, sum, 0) < 0)
+			    return DROP_CSUM_L3;
+
+			if (ct == CT_NEW) {
```
You set `ct` to `CT_NEW` on line 1379, so this condition is always true. What was the intent here? Why do we need to check the conntrack state in the first place?

Jiang1155 replied:
I should remove line 1379. The intention is to pass CT status into this function, then only create a nat entry for the new connections.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+
+			*dsr = false;
+			/* this will remove inner iph */
+			if (ctx_adjust_hroom(ctx, -20,
```
We should use `sizeof(struct iphdr)` here.

Jiang1155 replied:
ok

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+		podip = ip4->daddr;
+
+		if (ip4->protocol == IPPROTO_IPIP) {
+			if (ctx_load_bytes(ctx, ETH_HLEN + sizeof(struct iphdr),
```
We should check that the outer header isn't bigger than `sizeof(struct iphdr)`.

Jiang1155 replied:
Fixed

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+		ct = 0;
+		if (!revalidate_data(ctx, &data, &data_end, &ip4))
+			return DROP_INVALID;
+		*dsr = false;
```
This is not needed since `dsr` is already initialized to false.

Jiang1155 replied:
agree

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+					&iph_inner, sizeof(iph_inner)) < 0)
+				return DROP_INVALID;
+
+			*dsr = false;
```
This is not needed since `dsr` is already initialized to false.

Jiang1155 replied:
agree

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+				if (ctx_load_bytes(ctx, l4_off +2, &dport, 2) < 0)
+					return DROP_CT_INVALID_HDR;
+
+				if (snat_v4_create_dsr(ctx, vip, dport) < 0)
```
We should set `*dsr = true` before calling that function, no?

Jiang1155 replied:
I think you are right. Will fix it.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+					return DROP_INVALID;
+
+				l4_off = l3_off + ipv4_hdrlen(ip4);
+				if (ctx_load_bytes(ctx, l4_off +2, &dport, 2) < 0)
```
Let's not hardcode the offset here.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1360,11 +1360,73 @@ static __always_inline int dsr_set_opt4(struct __ctx_buff *ctx,
+
+#elif DSR_ENCAP_MODE == DSR_ENCAP_NONE
+	if (ct != CT_NEW)
+		return 0;
```
Why is this check needed?

Jiang1155 replied:
This check is needed to make old DSR option logic the same as before.  Before the patch, handle_dsr_v4 is only called when CT is new ( and reopened in later version ).

pchaigno commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -351,9 +351,6 @@ func initKubeProxyReplacementOptions() (bool, error) {
 			option.Config.LoadBalancerDSRDispatch == option.DSRDispatchIPIP {
-			if option.Config.DatapathMode != datapathOption.DatapathModeLBOnly {
-				return false, fmt.Errorf("DSR dispatch mode %s only supported for --%s=%s", option.Config.LoadBalancerDSRDispatch, option.DatapathMode, datapathOption.DatapathModeLBOnly)
-			}
```
Could you also update the documentation at https://docs.cilium.io/en/latest/gettingstarted/kubeproxy-free/#direct-server-return-dsr?

Jiang1155 replied:
updated

pchaigno commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -351,9 +351,6 @@ func initKubeProxyReplacementOptions() (bool, error) {
 			option.Config.LoadBalancerDSRDispatch == option.DSRDispatchIPIP {
-			if option.Config.DatapathMode != datapathOption.DatapathModeLBOnly {
-				return false, fmt.Errorf("DSR dispatch mode %s only supported for --%s=%s", option.Config.LoadBalancerDSRDispatch, option.DatapathMode, datapathOption.DatapathModeLBOnly)
-			}
```
We also need to extend bpf/Makefile's `LXC_OPTIONS` with new cases to cover `DSR_ENCAP_MODE == DSR_ENCAP_NONE` and `DSR_ENCAP_MODE == DSR_ENCAP_IPIP`. See `XDP_OPTIONS` for an example.

Jiang1155 replied:
In the latest version, I added some MACROs in XDP_OPTIONS in makefile. Not sure if that is good enough or not. For now, this new ipip mode requires XDP.

pchaigno reviewed (changes requested):
Thanks for your contribution! Being able to use IPIP with DSR without needing the LB-only mode would be great!

I reviewed this with @YutaroHayakawa. Please see our comments below. In addition to those, could you:
1. Update the release note text in the pull request description? That will be integrated to the v1.12 release notes so it needs to describe the user impact. One sentence is enough.
2. Add an end-to-end test. You can find an example of an existing end-to-end test for XDP+DSR at https://github.com/cilium/cilium/blob/bcb2db046b49dede753f91799552f31ec687f7e0/test/k8sT/Services.go#L1531-L1541.

Don't hesitate to ask on Slack if you need help with any of the suggested changes!

Jiang-BD committed:
Add dsrL4Translate option for k8s

Current cilium agent supports --bpf-lb-dsr-l4-xlate option. But
helm does not support this option yet. This patch added dsrL4Translate
to be used when installing with hem. This will be used in the
CI test too.

Signed-off-by: Jiang Wang <jiang.wang@bytedance.com>

Jiang-BD committed:
Add test for dsr ipipcni option

Signed-off-by: Jiang Wang <jiang.wang@bytedance.com>

pchaigno commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -497,6 +498,42 @@ mode would look as follows:
+
+.. _DSR mode with IPIP:
+
+Direct Server Return (DSR) with IPIP
```
Shouldn't this section be right after the ` Direct Server Return (DSR) with IP option` section?

Jiang1155 replied:
Sure. Will do.

pchaigno commented on bpf/Makefile:
```diff
@@ -208,8 +208,10 @@ XDP_OPTIONS = $(LB_OPTIONS) \
 	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_HOST_SERVICES_UDP:-DENABLE_HOST_SERVICES_TCP:-DENABLE_NODEPORT:-DENABLE_DSR:-DENABLE_DSR_HYBRID: \
 	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DDSR_ENCAP_MODE:-DDSR_ENCAP_NONE:-DDSR_ENCAP_IPIP=2 \
 	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DDSR_ENCAP_MODE:-DDSR_ENCAP_IPIP:-DDSR_ENCAP_NONE=2 \
+	-DENABLE_NODEPORT_ACCELERATION:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NODEPORT:-DENABLE_DSR:-DDSR_ENCAP_MODE:-DDSR_ENCAP_IPIP_CNI:-DDSR_ENCAP_NONE=3 \
```
Why 3?

pchaigno replied:
We should also update the configs under `bpf/complexity-tests/` to cover this new option. It's likely to have a complexity higher than the current DSR options.

Jiang1155 replied:
I will change that to 2. I was a little bit confused when I changing it to 3.  Will also update complexity-tests. I think just need to update bpf_lxc.txt file, right?

pchaigno commented on bpf/lib/common.h:
```diff
@@ -609,9 +609,11 @@ enum {
 	CB_POLICY,
 #define	CB_ADDR_V6_2		CB_POLICY	/* Alias, non-overlapping */
 #define	CB_BACKEND_ID		CB_POLICY	/* Alias, non-overlapping */
+#define CB_SRC_PORT		CB_POLICY	/* Alias, non-overlapping */
```
Watch out for the mix of tab and space indentations here.

Jiang1155 replied:
will do.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -1602,18 +1602,31 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, enum ct_status
 		int ret2;
 
-		ret2 = handle_dsr_v4(ctx, &dsr);
+		ret2 = handle_dsr_v4(ctx, &dsr, (enum ct_status)ret);
```
```suggestion
		ret2 = handle_dsr_v4(ctx, &dsr, *ct_status);
```

Jiang1155 replied:
fixed

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -1602,18 +1602,31 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, enum ct_status
 #endif /* ENABLE_PER_PACKET_LB && !DISABLE_LOOPBACK_LB */
 
 #ifdef ENABLE_NODEPORT
-	if (ret == CT_NEW || ret == CT_REOPENED) {
```
Why is this change required? Won't it impact other options besides the new one?

Jiang1155 replied:
I moved CT== new or reopened check inside the handle_dsr_v4() function, so it will not impact old DSR with IP option mode. 

But this is a good question, it reminds me that maybe I can optimize current code a little bit.
My original thought is that we will do IPIP encap for all tcp packets (not only syn packets as used in ip option mode).  But now I think maybe I don't need to do that. I can do something similar as IP option and only use IPIP encap for syn packets.  I will give it a try.

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -1602,18 +1602,31 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, enum ct_status
+		tuple.saddr = ip4->saddr;
+
+		ret = ct_lookup4(get_ct_map4(&tuple), &tuple, ctx, l4_off, CT_INGRESS, &ct_state,
+				 &monitor);
```
Why do we need to lookup the connection state again here when we already performed a lookup on line 1529?

Jiang1155 replied:
The tuple changed. The lookup at line 1529 is looking for outer tuples. After handle_dsr_v4 on line 1610, the nexthdr and saddr changed. If I don't do lookup here, it still works. But there will be duplicate ct entries for a single connection.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -523,6 +523,9 @@ int tail_nodeport_ipv6_dsr(struct __ctx_buff *ctx)
 	port = (__u16)ctx_load_meta(ctx, CB_PORT);
 	ret = dsr_set_ext6(ctx, ip6, &addr, port, &ohead);
+#elif DSR_ENCAP_MODE == DSR_ENCAP_IPIP_CNI
+	/*To do, add support for ipv6 */
```
That limitation should be documented.

Jiang1155 replied:
Sure. will do.

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -753,7 +756,11 @@ static __always_inline int nodeport_lb6(struct __ctx_buff *ctx,
 	svc = lb6_lookup_service(&key, false);
 	if (svc) {
-		const bool skip_l3_xlate = DSR_ENCAP_MODE == DSR_ENCAP_IPIP;
+#if DSR_ENCAP_MODE == DSR_ENCAP_IPIP
+		const bool skip_l3_xlate = true;
+#else
+		const bool skip_l3_xlate = false;
+#endif
```
```suggestion
		const bool skip_l3_xlate = DSR_ENCAP_MODE == DSR_ENCAP_IPIP;
```

Jiang1155 replied:
I first attempt is the same as you suggested. But I got following error and had a hard time to figure out why.

/work/go/src/github.com/cilium/cilium/bpf/lib/nodeport.h:766:48: error: use of undeclared identifier 'DSR_ENCAP_IPIP'
                const bool skip_l3_xlate = DSR_ENCAP_MODE == DSR_ENCAP_IPIP;

So I used current way to make the compiler happy.

pchaigno reviewed (changes requested):
I didn't review the core of the logic, but have a few other minor comments that may be worth addressing in the next iteration.

brb reviewed (changes requested):
Thanks for the PR! I think you need to mention in the docs and also implement a check in `pkg/service/service.go` to check that a backend port matches a service port, as with the IP-in-IP encapsulation we cannot have L4 different ports.

Jiang-BD committed:
Support dsr in ipip mode for ipv4 packets

Current Cilium only supports DSR by adding an customized IP option to the
headers. Many switches do not recognize this IP option. Therefore,
those packets will go through the control plane of network switches,
which introduces extra overhead. To have a better
performance, this patch adds support for DSR using IPIP. In DSR IPIP
mode, the service IP and port will be put into IP option of the inner
IP packet, The outer IP header does not have any IP options,
so the switches will forward the packets via a fast path.

In sum, this patch did following:
1. Add a new "ipipcni" option for --bpf-lb-dsr-dispatch. When set, the
packets are encapsulated with IPIP. The difference with existing "ipip"
option is that the service IP and port are put into the inner IP options,
and inner dst IP is the backend IP.
2. Allow and decap IPIP v4 packets in ingress. Remove outer ip header and only
keep inner ip header. Doing this by overwrite ip header.
3. Create an NAT entry for those packets.

Signed-off-by: Jiang Wang <jiang.wang@bytedance.com>
